### PR TITLE
Add clamp functions

### PR DESCRIPTION
**Pull request by Xaeroxe**
Implementation of clamp feature:

Tracking issue: https://github.com/rust-lang/rust/issues/44095
RFC: https://github.com/rust-lang/rfcs/pull/1961

### CODE REVIEW

File: src/libstd/f32.rs
Comment (scottmcm) [inline src/libstd/f32.rs]: Perhaps an example of NAN input giving NAN output?

File: src/libsyntax/feature_gate.rs
Comment (scottmcm) [inline src/libsyntax/feature_gate.rs]: No need for the feature gate in code for lib changes; the attribute is sufficient.

File: src/libstd/f32.rs
Comment (lukaramu) [inline src/libstd/f32.rs]: Should this be "is `NaN`" and not "equals `NaN`"? Nothing is equal (as in `==`) to `NaN`, not even `NaN` (if I recall correctly).

File: src/libstd/f64.rs
Comment (lukaramu) [inline src/libstd/f64.rs]: Same as above.

**Comment by carols10cents** [comment]
Thanks for the PR! We’ll periodically check in on it to make sure that @BurntSushi or someone else from the team reviews it soon.

File: src/libcore/cmp.rs
Comment (BurntSushi) [inline src/libcore/cmp.rs]: I think this should follow our doc formatting conventions, no? I think we usually have a panics header for documenting panic'ing conditions.

File: src/libcore/cmp.rs
Comment (BurntSushi) [inline src/libcore/cmp.rs]: Could you format this in a way that is consistent with the surrounding code? (Do we do rustmt on this stuff yet?)

File: src/libstd/f32.rs
Comment (BurntSushi) [inline src/libstd/f32.rs]: Needs panic header for here too.

File: src/libstd/f64.rs
Comment (BurntSushi) [inline src/libstd/f64.rs]: Needs panic header.

**Review by BurntSushi — CHANGES_REQUESTED** [review]
This mostly looks good, but have some nits! I guess this should also technically wait to merge until the RFC has been merged?

**Comment by Xaeroxe** [comment]
r? @BurntSushi 

Yeah this will have to wait for the RFC to be merged.  I mostly opened this early since the implementation was so straightforward.

**Review by BurntSushi — CHANGES_REQUESTED** [review]
Oh, and could you also add some unit tests for the panic conditions? Thanks!

**Comment by Xaeroxe** [comment]
@aturon @BurntSushi That didn't seem to work :/

**Comment by aturon** [comment]
@Xaeroxe it's on [the queue](https://buildbot2.rust-lang.org/homu/queue/rust), which is pretty backlogged at the moment.

File: src/libstd/f64.rs
Comment (StefanoD) [inline src/libstd/f64.rs]: I'm a Rust beginner, but I think, this should also work. I don't like the ```mut```.

```rust
if x < min { min }
if x > max { max }
else { self }
```

Comment (Xaeroxe) [reply src/libstd/f64.rs]: That was the original implementation.  This specific code pattern was chosen because it yields a highly efficient assembly implementation.  This was actually the source of quite a bit of conversation before we settled on this.

Comment (StefanoD) [reply src/libstd/f64.rs]: Oh, interesting and ```let x```?

Comment (Xaeroxe) [reply src/libstd/f64.rs]: Yeah, and that.

Comment (StefanoD) [reply src/libstd/f64.rs]: Holy caroly (this is actually no real word, but it rhymes)!

**Comment by pcwalton** [comment]
This broke Servo and Pathfinder. See https://github.com/servo/app_units/pull/37

**Pull request merged**
